modelserver-markers-theia/src/browser/diagnostic-manager.ts

```diff
@@ -0,0 +1,74 @@
+import { Diagnostic as EMFDiagnostic } from '@eclipse-emfcloud/modelserver-theia/lib/browser';
+import URI from '@theia/core/lib/common/uri';
+import { ProblemManager } from '@theia/markers/lib/browser';
+import { inject, injectable } from 'inversify';
+import { Diagnostic as LangServerDiagnostic } from 'vscode-languageserver-types';
+import { DiagnosticAdapter } from './diagnostic-adapter';
+
+@injectable()
+export class DiagnosticManager {
+
+    /** The problem manager to which we delegate */
+    @inject(ProblemManager)
+    protected readonly problemManager: ProblemManager;
+
+    /**
+     * Replaces the current markers for the given model URI with markers generated from the given diagnostic.
+     * @param modelUri the model URI concerned by diagnostic
+     * @param diagnostic the root diagnostic (children diagnostics will be collected if needed)
+     */
+    setDiagnostic(modelUri: URI, diagnostic: EMFDiagnostic): void {
+        const leaves: EMFDiagnostic[] = EMFDiagnostic.collectLeaves(diagnostic);
+        this.setDiagnosticLeaves(modelUri, leaves);
+    }
+
+    /**
+     * Replaces the current markers for the given model URI with markers generated from the given diagnostics.
+     * @param modelUri the model URI concerned by diagnostic
+     * @param diagnostics the diagnostics to take (children diagnostics will be collected if needed)
+     */
+    setDiagnostics(modelUri: URI, diagnostics: EMFDiagnostic[]): void {
+        const leavesPerDiagnostic: EMFDiagnostic[][] = diagnostics.map(d => EMFDiagnostic.collectLeaves(d));
+        const leaves: EMFDiagnostic[] = leavesPerDiagnostic.reduce((accumulator, values) => accumulator.concat(values), []);
+        this.setDiagnosticLeaves(modelUri, leaves);
+    }
+
+    /**
+     * Replaces the current markers for the given model URI with markers generated from the given leaf diagnostics.
+     * @param modelUri the model URI concerned by diagnostic
+     * @param diagnostics the leaf diagnostics to take as is
+     */
+    protected setDiagnosticLeaves(modelUri: URI, diagnostics: EMFDiagnostic[]): void {
+        // first clean old markers
+        this.problemManager.cleanAllMarkers(modelUri);
+        // then convert diagnostics
+        const convertedDiagnosticsPerElementId: Map<string, LangServerDiagnostic[]> = new Map();
+        diagnostics.forEach(d => {
+            const converted = DiagnosticAdapter.emfToLangServerDiagnostic(d);
+            // check not undefined...
+            if (LangServerDiagnostic.is(converted)) {
+                if (!convertedDiagnosticsPerElementId.has(d.id)) {
+                    convertedDiagnosticsPerElementId.set(d.id, []);
+                }
+                convertedDiagnosticsPerElementId.get(d.id)!.push(converted);
+            }
+        });
+        // and log them by element id as owner
+        for (const [id, converteds] of convertedDiagnosticsPerElementId) {
```

Comment:
Is `converteds` a typo?

Reply:
Ah, now I see how it should be read. No, you can leave it as is if you want. Don't wanna be too nit-picky and I really was just wondering about it ;-)